content/name-based-xml.md

```diff
@@ -0,0 +1,421 @@
+* `noPrefix`
+* `prefixes`
+* `uris`
+* `literal`
+
+All above methods except `literal` should return a builder, and `literal` will turn one or more builders into an XML object / or an XML node list.
+
+An XML library user can switch different implementations by importing different `xml` packages or objects. `scala.xml` is used by default when no explicit import is present.
+
+In a schema-aware XML library like Binding.scala, its `elements` , `attributes` , `processInstructions` and `entities` methods should return factory objects that contain all the definitions of available tag names and attribute names. An XML library user can provide additional tag names and attribute names in user-defined implicit classes for `tags` and `attributes` .
+
+In a schema-less XML library like `scala-xml` , its `elements` , `attributes` , `processInstructions` and `entities` should return builders that extend [scala.Dynamic](https://www.scala-lang.org/api/current/scala/Dynamic.html) in order to handle tag names and attribute names in `selectDynamic` or `applyDynamic` .
+
+Those XML libraries can be extended with the help of standard XML namespace bindings. A plug-in author can create `implicit class` for `xml.uris` to introduce foreign elements embedded in existing XML literals.
+
+[html.scala](https://github.com/GlasslabGames/html.scala) is an XML library vendor for creating reactive HTML templates.
+
+## Drawbacks
+
+### Name clash
+
+`<toString/>` or `<foo equals="bar"/>` will not compile due to name clash to `Any.toString` and `Any.equals` .
+
+* Compilation error is the desired behavior in a schema-aware XML library as long as `toString` is not a valid name in the schema. Fortunately, unlike JSX, `<div class="foo"></div>` should compile because `class` is a valid method name.
+* A schema-less XML library user should instead explicit construct `<toString/>` in Scala code, e.g. `new Elem("toString")` .
+
+## Alternative approach
+
+XML initialization can be implemented in a special string interpolation as `xml"<x/>"`, which can be implemented in a macro library. [scala-xml-quote](https://github.com/densh/scala-xml-quote) is an implementation in this approach. The pros and cons of these approaches are listed in the following table:
+
+||symbol-based XML literals in Scala 2.12|name-based XML literals in this proposal|`xml` string interpolation|
+| --- | --- | --- | --- |
+|XML is parsed by ...|compiler|compiler|library, IDE, and other code browsers including Github, Jekyll (if syntax highlighting is wanted)|
+|Is third-party schema-less XML library supported?|No, unless using white box macros|Yes|Yes|
+|Is third-party schema-aware XML library supported?|No, unless using white box macros|Yes|No, unless using white box macros|
+|How to highlight XML syntax?|By regular highlighter grammars|By regular highlighter grammars|By special parsing rule for string content|
+|Can presentation compiler perform code completion for schema-aware XML literals?|No|Yes|No|
```

Comment:
Another point of comparison, where probably this proposal comes out on top, is IDE support for navigation (go to definition, find references, etc..).

A name-based approach would allow things like "find all the usages of `<div />`" pretty much automatically (provided that dotc outputs all the relevant information to the SemanticDB), whereas using a string literal would require non-trivial work (to parse and index all the XML literals, and even that would be string-based, so it could only guess matches across literals based on the tag name)